druid-shell/src/platform/mac/window.rs

```diff
@@ -725,23 +735,21 @@ extern "C" fn view_will_draw(this: &mut Object, _: Sel) {
 
 extern "C" fn draw_rect(this: &mut Object, _: Sel, dirtyRect: NSRect) {
     unsafe {
+        let view_state: *mut c_void = *this.get_ivar("viewState");
+        let view_state = &mut *(view_state as *mut ViewState);
+
+        let invalid: Region = mem::replace(
+            &mut view_state.invalid_region.lock().unwrap(),
+            Region::EMPTY,
+        );
+
         let context: id = msg_send![class![NSGraphicsContext], currentContext];
         //FIXME: when core_graphics is at 0.20, we should be able to use
         //core_graphics::sys::CGContextRef as our pointer type.
         let cgcontext_ptr: *mut <CGContextRef as ForeignTypeRef>::CType =
             msg_send![context, CGContext];
         let cgcontext_ref = CGContextRef::from_ptr_mut(cgcontext_ptr);
 
-        // FIXME: use the actual invalid region instead of just this bounding box.
-        // https://developer.apple.com/documentation/appkit/nsview/1483772-getrectsbeingdrawn?language=objc
```

Comment:
Did you try using `getRectsBeingDrawn` instead of `dirtyRect`?

Reply:
Yes, `getRectsBeingDrawn` were returning one rect in my tests, same as `dirtyRect`.

Reply:
Hooray for accurate API docs :shrug: